src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+contract ALE is Ownable, ReentrancyGuard {
+    error CollateralNotSet();
+    error MarketNotSetForCollateral(address collateral);
+    error SwapFailed();
+    error DOLAInvalidBorrow();
+    error DOLAInvalidRepay();
+    error InvalidProxyAddress();
+
+    // 0x ExchangeProxy address.
+    // See https://docs.0x.org/developer-resources/contract-addresses
+    address payable public exchangeProxy;
+
+    IDola public immutable dola;
+
+    struct Market {
+        address market;
+        address helper;
+        address collateral;
+    }
+
+    struct Permit {
+        uint256 deadline;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    // Mapping of sellToken/buyToken to Market structs
+    // NOTE: in normal cases sellToken/buyToken is the collateral token,
+    // in other cases it could be different (eg. st-yCRV is collateral, yCRV is the token to be swapped from/to DOLA)
+    // That's why we need to keep track of collateral in market struct
+    mapping(address => Market) public markets;
+
+    constructor(address _dola, address _exchangeProxy) Ownable(msg.sender) {
+        dola = IDola(_dola);
+        exchangeProxy = payable(address(_exchangeProxy));
+    }
+
+    function setExchangeProxy(address _exchangeProxy) external onlyOwner {
+        if (_exchangeProxy == address(0)) revert InvalidProxyAddress();
+        exchangeProxy = payable(_exchangeProxy);
+    }
+
+    function setMarket(
+        address _market,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].market = _market;
+    }
+
+    function setMarketHelper(
+        address _helper,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].helper = _helper;
+    }
+
+    function setMarketCollateral(
+        address _collateral,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].collateral = _collateral;
+    }
+
+    function leveragePosition(
+        // Amount of DOLA to borrow
+        uint256 _value,
+        // `buyTokenAddress` field from the API response.
+        address _buyTokenAddress,
+        // The `allowanceTarget` field from the API response.
+        address _spender,
+        // The `data` field from the API response.
+        bytes calldata _swapCallData,
+        // Permit data
+        Permit calldata _permit,
+        // Optional helper data in case the collateral needs to be transformed
+        bytes calldata _helperData
+    ) external payable nonReentrant {
+        if (markets[_buyTokenAddress].market == address(0))
+            revert MarketNotSetForCollateral(_buyTokenAddress);
+
+        IMarket market = IMarket(markets[_buyTokenAddress].market);
+
+        // Mint and approve
+        dola.mint(address(this), _value);
+        dola.approve(_spender, _value);
+
+        IERC20 buyToken = IERC20(_buyTokenAddress);
+        uint256 collateralBalBefore = buyToken.balanceOf(address(this));
```

Comment:
Is there any reason to think there will be a `buyToken` balance in the contract?

Reply:
The idea was to check that we did actually get the collateral from the swap and not because it was directly sent to the contract, in theory it was to avoid miscalculation but after reviewing the 0x api and our logic, i think it can be removed, fixed in commit 829bf77